docs/design.md

```diff
@@ -0,0 +1,284 @@
+you can mix and match versioning strategies with language support.
+
+## Language support
+
+`release-please` is highly extendable to support new languages and package managers. We
+currently support 20+ different `Strategy` types many of which are community created and
+supported.
+
+### Strategy
+
+We define a [`Strategy`][strategy] interface which abstracts the notion of what files to
+update when proposing the next release version. In the most basic case
+([`Simple`][simple-strategy]), we do not update any source files except the `CHANGELOG.md`.
+
+**Contributor note**: Implementation-wise, most strategies inherit from the
+[`BaseStrategy`](base-strategy). This is not necessary, but it handles most of the common
+behavior. If you choose to extend `BaseStrategy`, you only need to implement a single
+`buildUpdates()` method (which files need to be updated).
+
+**Contributor note**: If you implement a new `Strategy`, be sure to add a new corresponding
+test to ensure we don't break it in the future.
+
+### Updating file contents
+
+The most common customation a `Strategy` makes is determining which standard files need to be
+updated. For example, in a `nodejs` library, you will want to update the `version` entry in
+your library's `package.json` (and `package-lock.json` if it exists).
+
+We represent a file update via the [`Update`][update] interface. An `Update` contains the
+path to the file needing an update, whether or not to create the file if it does not exist,
+and how to update the file. The [`Updater`][update] interface is an abstraction that is
+actually responsible for updating the contents of a file. An `Updater` implementation
+generates updated content given the original file contents and a new version (or versions)
+to update within that file.
+
+**Contributor note**: If you implement a new `Updater`, be sure to add a new corresponding
+test to ensure we don't break it in the future.
+
+## Changelog/release notes
+
+We define a [`ChangelogNotes`][changelog-notes] interface which abstracts the notion of how
+to build a `CHANGELOG.md` entry given a list of commits. The default implementation
+([`DefaultChangelogNotes`][default-changelog-notes]), uses the
+`conventional-changelog-writer` library to generate a standardized release notes based on
+the conventionalcommits.org specification.
+
+We also have a second implementation that uses the GitHub changelog generator API.
+
+## Release pull request
+
+`release-please` operates without a database of information and so it relies on GitHub as
+the source of it's information. In that way, the release pull request is heavily formatted
+and its structure is load-bearing.
+
+### Branch name
+
+The name of the HEAD branch that `release-please` creates its pull request from contains
+important information that `release-please` needs.
+
+As such, we implement a helper [`BranchName` class][branch-name] that encapsulates that
+data.
+
+The branch name is not customizable at this time.
```

Comment:
Earlier, there is "Usually default branch but can be configured to any".
I felt these statements are contradicting. Can you add clarification?

Reply:
This is the name of the HEAD branch. In GitHub terms for a pull request there is the head branch (contains the code to be merged) and the base branch (the target branch). The target branch is configurable, the name of the branch release-please pushes code to is not.

Reply:
Updated to clarify that the "HEAD branch" is not customizable at this time.